draft-yasskin-http-origin-signed-responses.md

```diff
@@ -518,17 +505,14 @@ to retrieve an updated OCSP from the original server.
       Exchange 1 ” instead. This ensures that signers can predict how their
       signatures will be used.
    1. A single 0 byte which serves as a separator.
-   1. The bytes of the canonical CBOR serialization ({{canonical-cbor}}) of a
-      CBOR map mapping:
-      1. If `cert-sha256` is set:
-         1. The text string "cert-sha256" to the byte string value of
-            `cert-sha256`.
-      1. The text string "validity-url" to the byte string value of
-         `validity-url`.
-      1. The text string "date" to the integer value of `date`.
-      1. The text string "expires" to the integer value of `expires`.
-      1. The text string "headers" to the CBOR representation
-         ({{cbor-representation}}) of `exchange`'s headers.
+   1. If `cert-sha256` is set, a byte holding the value 32 followed by the 32
+      bytes of the value of `cert-sha256`. Otherwise a 0 byte.
+   1. The 8-byte big-endian encoding of the length in bytes of `validity-url`,
+      followed by the bytes of `validity-url`.
+   1. The 8-byte big-endian encoding of `date`.
+   1. The 8-byte big-endian encoding of `expires`.
+   1. The 8-byte big-endian encoding of the length in bytes of `headers`,
+      followed by the bytes of `headers`.
```

Comment:
I think it's fine for this part to be CBOR if you all want it to be CBOR.

This format's fine too, though if you do that, I suggest using BoringSSL's `CBB` functions in the implementation (we don't currently have `CBB_add_u64{,_length_prefixed}` since TLS doesn't use those, but they'd be easy to add). TLS also has a struct syntax you could borrow. (An 8 byte length prefix is also kind of impressively large for both URL and headers. :-) I don't think any implementation would tolerate either field to exceed, say, a 3 byte length. But 8 bytes is cool too if you prefer.)

To the CBOR point, I think I didn't express the concern clearly. My bad. It's not that you shouldn't trust the CBOR encoder itself. Rather, let's look at the pipeline of conversions we've got here.

We convert the byte representation of the signed exchange, to a CBOR AST (CBORValue), to some parsed signed exchange struct (SignedExchangeEnvelope), to the net/loading stack's representation (HttpResponseHeaders). Along the way, to verify the signature, we convert the SignedExchangeEnvelope *back* to the CBOR AST, then *back* to bytes, and *then* verify it.

That is:

```
bytes -> CBORValue -> SignedExchangeEnvelope -> HttpResponseHeaders -...-> web stack
                                 |
                                 v
                             CBORValue
                                 |
                                 v
                               bytes
                                 |
                                 v
                         (verify signature)
```

We want the "verify signature" step to cover what was ultimately sent to the rest of the stack. That's the point of this whole process. We wish to know that everything downstream of the verification only sees content attested by the signature.

Suppose there's a logic bug in one of the right arrows. Maybe on certain malformed or weird input, we output something that can be massaged into an XSS. That's bad, but the signature mitigates a lot of potential attack vectors. The origin isn't going to sign the bad input, so as long as the verify step covered everything, an attacker can't use that bug to inject bad input into the origin. *The requirement on right arrows is the depend only on their inputs.*

However, suppose there is a logic bug in one of the down arrows. Maybe we forgot to copy some new field from the SignedExchangeEnvelope into the CBORValue. That's a problem. An attacker can put the new field in, the signature would check out, and all the right arrows would think this new field was covered, when it really wasn't. *The requirement on down arrows is they preserve all information.*

Note that, of these two requirements, one is much more straightforward than the other. It's hard to accidentally check a random global variable somewhere. :-) It's much easier to accidentally forget to encode something.

Thus the goal is to minimize the complexity in the verify step. The trick I like most is to find the simplest representation and verify that. An array of bytes is a good candidate. No one is going to extend that. All the complex structure is carried within, and the verify step is blind to it all.

Now, you can't always manage this. In the case of this logic, you are not just bytes. It's bytes plus some metadata decoded out of headers (cert-sha256, validity-url, etc). And, yeah, that means you need to go encode that metadata in some way. Maybe CBOR, maybe DER, or maybe something ad hoc. And, yeah, that's more down arrows. It would be more robust if you didn't have to do that, but there are also benefits to having the signature metadata outside of the payload, so the tradeoff is probably to leave those as is. Hopefully we don't have to extend those very often and the right arrows probably discard that information pretty quickly.

The rest of the payload, on the other hand, seems preferable to keep as bytes. Maybe someone will try to put trailers in there someday, I dunno.